### PR TITLE
Replace Overflow-Prone Operations in Staking with Saturating Arithmetic

### PR DESCRIPTION
**Pull request by Kianenigma**
A minimal and temporary fix until a better approach is introduced --for good.

### CODE REVIEW

File: srml/staking/src/phragmen.rs [hunk 1]
```diff
@@ -330,25 +333,29 @@ pub fn equalise<T: Trait + 'static>(
...
 		}
-		cumulative_stake += stake;
+		cumulative_stake = cumulative_stake.saturating_add(stake);
 	});
 
 	let last_stake = elected_candidates[elected_edges[last_index].elected_idx].backing_stake;
 	let split_ways = last_index + 1;
-	let excess = nominator.budget + cumulative_stake - last_stake * <BalanceOf<T>>::sa(split_ways as u64);
+	let excess = nominator.budget
+	.saturating_add(cumulative_stake)
```
Comment (thiolliere) [inline srml/staking/src/phragmen.rs]: I think the format guidelines is to ident one more time here

File: srml/staking/src/phragmen.rs [hunk 2]
```diff
@@ -330,25 +333,29 @@ pub fn equalise<T: Trait + 'static>(
...
+	.saturating_sub(
+		last_stake.checked_mul(&<BalanceOf<T>>::sa(split_ways as u64))
+		.unwrap_or(<BalanceOf<T>>::max_value())
```
Comment (thiolliere) [inline srml/staking/src/phragmen.rs]: I think the format guidelines is to ident one more time here

File: srml/staking/src/tests.rs [hunk 3]
```diff
@@ -1885,3 +1885,105 @@ fn phragmen_chooses_correct_validators() {
...
+
+		System::set_block_number(2);
+		Session::check_rotate_session(System::block_number());
+
+		assert_eq!(Session::validators(), vec![4, 2]);
+	})
+}
+
+#[test]
+fn phragmen_should_not_overflow_ultimate() {
```
Comment (thiolliere) [inline srml/staking/src/tests.rs]: do you think this test is a superset of the two above ?
if one of the above implies this one fails then I'm mixed about putting them as they are already covered. But still ok though

File: srml/staking/src/phragmen.rs [hunk 4]
```diff
@@ -330,25 +333,29 @@ pub fn equalise<T: Trait + 'static>(
...
 			return
 		}
-		cumulative_stake += stake;
+		cumulative_stake = cumulative_stake.saturating_add(stake);
 	});
 
 	let last_stake = elected_candidates[elected_edges[last_index].elected_idx].backing_stake;
 	let split_ways = last_index + 1;
-	let excess = nominator.budget + cumulative_stake - last_stake * <BalanceOf<T>>::sa(split_ways as u64);
+	let excess = nominator.budget
```
Comment (thiolliere) [inline srml/staking/src/phragmen.rs]: note: also this change is safer it makes things less readable

File: srml/staking/src/tests.rs (hunk 3 above)
Comment (Kianenigma) [reply srml/staking/src/tests.rs]: It _kind of is_; yet it is not intended to cover all cases. These tests are the basics of preventing of overflow and they are not intended to cover all cases. The main reason for that being the fact that the tests for Staking module are already quite a lot and while this approach is temporary, I didn't want to bloat the test file even more.

And the reason that all three of them are needed at least: The relationship between nominator's stake and validator's stake is such that if one is maximum w.r.t. `Perquintill` (e.g. `u64::max`), then some loads that are later on multiplied by the sake value end up being `1/QUINTILLION`. So we end up with something like `u64::max * 1` which is not good to demonstrate all overflow cases. Hence, it is needed for either to be the max and the other one to not be + the case was both are max.

**Comment by Kianenigma** [comment]
About division. I indeed have tought of it but the reason for not even using `checked_x` operations in favor of readability is: we have three divisions: 

```
let temp = n.budget.as_().saturating_mul(*n.load) / c.approval_stake.as_();
```
where `approval_stake` is filtered to be positive. It also implies that that `c.score` is positive, hence `n.load` being positive. This solves the next one:

```
e.backing_stake = <BalanceOf<T>>::sa(n.budget.as_().saturating_mul(*e.load) / *n.load);
```

Finally: 

```
e.backing_stake = (excess / <BalanceOf<T>>::sa(split_ways as u64)).saturating_add(last_stake) - c.backing_stake;
```

where `split_ways = last_index + 1;` where `last_index` is greater or equal to zero.

(I hope I am correct about all of this. Will double check it as we proceed again).

**Comment by DemiMarie-parity** [comment]
Should we just use a bignum library?

**Comment by Kianenigma** [comment]
> Should we just use a bignum library?

Definitely an option. Never brought this up with you @gavofyork. Your opinion?

**Comment by gavofyork** [comment]
Any code running on-chain needs to be lean to minimise the risk of a heavy-block attack (when a single block is forced to be so difficult to compute that is messes with the consensus system).

**Pull request merged**